panelapp/panels/models/historical_snapshot.py

```diff
@@ -0,0 +1,382 @@
+##
+## Copyright (c) 2016-2019 Genomics England Ltd.
+##
+## This file is part of PanelApp
+## (see https://panelapp.genomicsengland.co.uk).
+##
+## Licensed to the Apache Software Foundation (ASF) under one
+## or more contributor license agreements.  See the NOTICE file
+## distributed with this work for additional information
+## regarding copyright ownership.  The ASF licenses this file
+## to you under the Apache License, Version 2.0 (the
+## "License"); you may not use this file except in compliance
+## with the License.  You may obtain a copy of the License at
+##
+##   http://www.apache.org/licenses/LICENSE-2.0
+##
+## Unless required by applicable law or agreed to in writing,
+## software distributed under the License is distributed on an
+## "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+## KIND, either express or implied.  See the License for the
+## specific language governing permissions and limitations
+## under the License.
+##
+import csv
+from django.contrib.postgres.fields import JSONField
+from django.db import models
+from django.http import HttpResponse
+
+
+from .genepanel import GenePanel
+from api.v1.serializers import PanelSerializer
+from panels.utils import remove_non_ascii
+from webservices.utils import make_null, convert_moi, convert_gel_status
+import panelapp
+
+
+class HistoricalSnapshot(models.Model):
+    panel = models.ForeignKey(GenePanel, on_delete=models.PROTECT)
+    major_version = models.IntegerField(default=0, db_index=True)
+    minor_version = models.IntegerField(default=0, db_index=True)
+    reason = models.CharField(null=True, max_length=100)
+    schema_version = models.CharField(max_length=100)  # JSON schema version
+    data = JSONField()
+
+    def to_tsv(self):
```

Comment:
could you refactor this, so each row writer is a separate function? it will be easier to test. Maybe it's a good idea to move these functions to a separate file too.